File: MiniDOMBase/DMD/Anagrafica/Fonti/FontiDefaultProvider.vb
Imports minidom
Imports minidom.Sistema
Imports minidom.Databases



Partial Public Class Anagrafica
 
    ''' <summary>
    ''' Provider predefinito delle fonti
    ''' </summary>
    ''' <remarks></remarks>
    Public NotInheritable Class FontiDefaultProvider
        Implements IFonteProvider

        Public Sub New()
            DMDObject.IncreaseCounter(Me)
        End Sub

        Public Function GetItemsAsArray(ByVal tipo As String, Optional ByVal onlyValid As Boolean = True) As IFonte() Implements IFonteProvider.GetItemsAsArray
            tipo = Trim(tipo)
            Dim ret As New CCollection(Of IFonte)
            'Dim cursor As New CFontiCursor
            'cursor.Stato.Value = ObjectStatus.OBJECT_VALID
            'cursor.OnlyValid = onlyValid
            'cursor.Nome.SortOrder = SortEnum.SORT_ASC
            'If (nome <> vbNullString) Then cursor.Tipo.Value = nome
            'While Not cursor.EOF
            '    ret.Add(cursor.Item)
            '    cursor.MoveNext()
            'End While
            'cursor.Reset()
            Dim items As CCollection = Anagrafica.Fonti.LoadAll

            For i As Integer = 0 To items.Count - 1
                Dim f As CFonte = items(i)
                If (tipo = "" OrElse Strings.Compare(f.Tipo, tipo) = 0) Then ret.Add(f)
            Next
            Return ret.ToArray
        End Function



        Public Function GetItemById(ByVal tipo As String, id As Integer) As IFonte Implements IFonteProvider.GetItemById
            Return Fonti.GetItemById(id)
        End Function


        Public Function GetItemById(ByVal tipo As String, ByVal name As String) As IFonte Implements IFonteProvider.GetItemByName
            Return Fonti.GetItemByName(name)
        End Function

        Public Function GetSupportedNames() As String() Implements IFonteProvider.GetSupportedNames

            Dim items As New System.Collections.Hashtable
            For Each f As CFonte In Anagrafica.Fonti.LoadAll
                Dim t As String = f.Tipo
                If (items.ContainsKey(t) = False) Then items.Add(t, t)
            Next

            Dim ret() As String = Nothing
            If (items.Count > 0) Then
                ReDim ret(items.Count - 1)
                items.Keys.CopyTo(ret, 0)
                System.Array.Sort(ret)
            End If
            items = Nothing

            Return ret
        End Function

        Protected Overrides Sub Finalize()
            MyBase.Finalize()
            DMDObject.DecreaseCounter(Me)
        End Sub
    End Class



End Class

File: MiniDOMBase/DMD/Anagrafica/Fonti/Providers/FontiDefaultProvider.cs
using DMD.Databases.Collections;
using System;
using System.Collections;
using System.Collections.Generic;

namespace minidom
{
    public partial class Anagrafica
    {

        /// <summary>
        /// Provider predefinito delle fonti
        /// </summary>
        /// <remarks></remarks>
        public sealed class FontiDefaultProvider 
            : IFonteProvider
        {

            /// <summary>
            /// Costruttore
            /// </summary>
            public FontiDefaultProvider()
            {
                ////DMDObject.IncreaseCounter(this);
            }

            /// <summary>
            /// Restituisce l'array delle fonti di tipo specifico
            /// </summary>
            /// <param name="tipo"></param>
            /// <param name="onlyValid"></param>
            /// <returns></returns>
            public IFonte[] GetItemsAsArray(string tipo, bool onlyValid = true)
            {
                tipo = DMD.Strings.Trim(tipo);

                var ret = new List<IFonte>();                 
                foreach(var f in minidom.Anagrafica.Fonti.LoadAll())
                {
                    if (string.IsNullOrEmpty(tipo) || DMD.Strings.EQ(f.Tipo, tipo, true))
                        ret.Add(f);
                }

                return ret.ToArray();
            }

            /// <summary>
            /// Restituisce la fonte in base all'id
            /// </summary>
            /// <param name="tipo"></param>
            /// <param name="id"></param>
            /// <returns></returns>
            public IFonte GetItemById(string tipo, int id)
            {
                return minidom.Anagrafica.Fonti.GetItemById(id);
            }

            /// <summary>
            /// Restituisce la fonte in base al nome
            /// </summary>
            /// <param name="tipo"></param>
            /// <param name="name"></param>
            /// <returns></returns>
            public IFonte GetItemByName(string tipo, string name)
            {
                return minidom.Anagrafica.Fonti.GetItemByName(name);
            }

            /// <summary>
            /// Restituisce i tipi di fonte supportati
            /// </summary>
            /// <returns></returns>
            public string[] GetSupportedNames()
            {
                var items = new Dictionary<string, string>();
                foreach (var f in minidom.Anagrafica.Fonti.LoadAll())
                {
                    string t = f.Tipo;
                    if (items.ContainsKey(t) == false)
                        items.Add(t, t);
                }

                string[] ret = null;
                if (items.Count > 0)
                {
                    ret = new string[items.Count];
                    items.Keys.CopyTo(ret, 0);
                    Array.Sort(ret);
                }

                items = null;
                return ret;
            }

            /// <summary>
            /// Distruttore
            /// </summary>
            ~FontiDefaultProvider()
            {
                ////DMDObject.DecreaseCounter(this);
            }
        }
    }
}